File: minimum-window-substring.rs
impl Solution {
  fn c2i(c: char) -> usize {
    if c.is_ascii_lowercase() { c as usize - 'a' as usize }
    else { c as usize - 'A' as usize + 26 }
  }

  pub fn min_window(s: String, t: String) -> String {
    let (mut sc, mut tc) = (vec![0; 52], vec![0; 52]);
    for c in t.chars() { tc[Self::c2i(c)] += 1; }

    let (mut r, mut cnt, n) = (0, 0, s.len());
    for i in 0..52 {
      if tc[i] == 0 { cnt += 1; }
    }

    let s = s.chars().collect::<Vec<char>>();
    let (mut ans, mut lpos) = (0x3f3f3f3f, 0);
    for l in 0..n {
      while r < n && cnt < 52 {
        let i = Self::c2i(s[r]);
        sc[i] += 1;
        if sc[i] == tc[i] { cnt += 1; }
        r += 1;
      }

      if cnt == 52 {
        if r - l < ans {
          ans = r - l;
          lpos = l;
        }
      }

      let i = Self::c2i(s[l]);
      if sc[i] == tc[i] { cnt -= 1; }
      sc[i] -= 1;
    }

    if ans == 0x3f3f3f3f { "".to_string() }
    else { s[lpos..lpos + ans].iter().collect() }
  }
}

File: m/minimum-window-substring.cc
class Solution {
public:
  string minWindow(string s, string t) {
    int st[128]{}, tt[128]{};
    for (char c : t) tt[c]++;

    auto check = [&]() {
      for (int i = 0; i < 128; ++i) {
        if (st[i] < tt[i]) return false;
      }
      return true;
    };

    int N = s.size(), r = 0, sl = -1, sr = N;
    for (int l = 0; l < N; ++l) {
      while (r < N && !check()) ++st[s[r++]];
      if (check() && r - l < sr - sl) {
        sl = l;
        sr = r;
      }
      --st[s[l]];
    }

    if (sl == -1) return "";
    return s.substr(sl, sr - sl);
  }
};
